New resource: `azurerm_redhatopenshift_cluster`

fedeoliv opened the pull request:
## azurerm_redhatopenshift_cluster
This PR introduces an Azure Red Hat OpenShift (ARO) provider targeting the issue [#3614](https://github.com/hashicorp/terraform-provider-azurerm/issues/3614).

## Known limitations

~~Currently the worker profile is immutable on the ARO API, more details [here](https://docs.microsoft.com/en-us/rest/api/openshift/open-shift-clusters/update#workerprofile). In practice, it means that:~~

~~* Any change on the profile (e.g. nodes count, disk size) will require a cluster recreation.~~
* Using `oc` to scale the `MachineSet` may cause problems as the cluster state may not be represented in Terraform state.

> Update: Immutable worker profile was associated with the API `2020-04-30` version. Not applicable anymore for the `2022-04-01` version.

datianshi commented:
@fedeoliv @tombuildsstuff  Just wonder what is the status of this PR. seems this is way off the main branch. Just wondering what was the reason this was not merged initially? I am interested in this feature as well as contributing.

katbyte commented:
@datianshi - i believe it was because the worker profile is immutable so after terraform creates the resource there would be no good/easy way to handle any updates made to the worker profile.

datianshi commented:
@katbyte  Thanks for the reply. It makes sense. But I still think it has value just for provisioning. It is the nature of the resource. Up to the end-user to know the risks of changing/updating. End-user can use lifecycle management to prevent destroy as well.

jgardner04 commented:
@katbyte does that mean this is not going to be added until RH changes the way RHO is deployed?

fedeoliv commented:
I agree with you @datianshi, I believe it's valuable providing the option of deploying ARO through Terraform even with this current limitation.

fedeoliv commented:
Apparently the new RHO version on the Azure SDK supports mutable worker profile attributes, see the `2022-04-01` version [here](https://github.com/Azure/azure-sdk-for-go/blob/main/services/redhatopenshift/mgmt/2022-04-01/redhatopenshift/models.go#L882). The previous `2020-04-30` version had immutable properties (details for comparison [here](https://github.com/Azure/azure-sdk-for-go/blob/main/services/redhatopenshift/mgmt/2020-04-30/redhatopenshift/models.go#L829)).

I'll update the resource to use this new version to validate if that works and make the required adjustments.

fedeoliv commented:
Awesome @datianshi, thanks for sharing this one! I'm working on updating the schema, CRUD methods and integration tests to match this new version. I'll keep you posted 😊.

fedeoliv commented:
Hey @tombuildsstuff @katbyte @datianshi the PR is ready to be reviewed again 😊. I've updated the code and documentation to match the `2022-04-01` schema and introduced new acceptance tests for nodes encryption at host and FIPS cryptographic modules.

fedeoliv commented:
> what stops this branch from being approved and merged?

@tombuildsstuff wondering if you have any update on this review?

piomin commented:
@tombuildsstuff could you merge it?

katbyte reviewed (changes requested):
Thanks for the PR @fedeoliv - i've given in a quick look and left some comments inline. while running the tests i encountered this error:

```
Test Failed

------- Stdout: -------
=== RUN   TestAccOpenShiftCluster_private
=== PAUSE TestAccOpenShiftCluster_private
=== CONT  TestAccOpenShiftCluster_private
    testcase.go:110: Step 1/1 error: Error running apply: exit status 1
        
        Error: creating Red Hat OpenShift Cluster "acctestaro220812175256421235" (Resource Group "acctestRG-aro-220812175256421235"): redhatopenshift.OpenShiftClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="DuplicateClientID" Message="The provided client ID '*******' is already in use by a cluster."
        
          with azurerm_redhatopenshift_cluster.test,
          on terraform_plugin_test.tf line 38, in resource "azurerm_redhatopenshift_cluster" "test":
          38: resource "azurerm_redhatopenshift_cluster" "test" {
        
--- FAIL: TestAccOpenShiftCluster_private (207.42s)
```

not sure what client id it is referring to?

fedeoliv commented on internal/services/redhatopenshift/redhatopenshift_cluster_resource.go:
```diff
@@ -0,0 +1,826 @@
+				},
+			},
+
+			"master_profile": {
```
Thanks @katbyte! I replaced all occurrences of `master` prefixes on the ARO resource and its corresponding documentation.

fedeoliv commented:
> not sure what client id it is referring to?

@katbyte looking at the error message apparently we can't associate a service principal with multiple clusters (1-1 relationship). The options I have in mind is running the acceptance tests sequentially, or somehow generate new service principals for each test. Any suggestion?

katbyte commented:
@fedeoliv - we can run them sequentially, also if that is a service limitation we should document it in the docs

fedeoliv commented:
> @fedeoliv - we can run them sequentially, also if that is a service limitation we should document it in the docs

Great, I updated the ARO documentation with a note about this limitation.

katbyte commented:
@fedeoliv - we still need to change the tests to run sequentially

katbyte commented:
<img width="1635" alt="image" src="https://user-images.githubusercontent.com/1638467/187278559-ce2afc1e-fbd3-44db-86b3-b403fc3fccc0.png">

tombuildsstuff commented:
@fedeoliv apologies for the contradiction - we can we update this to use a new Service Principal created as a part of each test - we can use the AzureAD provider to do that - see [this test](https://github.com/hashicorp/terraform-provider-azurerm/blob/a8895c8283f1722d6a0575543f0d5e66eb589f52/internal/services/mssql/mssql_managed_instance_active_directory_administrator_resource_test.go#L65-L83) as an example - which'd mean we can run these in parallel and avoid this limitation.

david-yu commented:
Hi @fedeoliv Thanks for your work on this! I've also have been behind the scenes prodding the right folks to review this PR since we also use OpenShift for Consul K8s testing. Looking forward to see this get merged!

mzarglis commented:
Any ETA on this? @fedeoliv Amazing work this has been a long time coming.

katbyte commented:
Hey @fedeoliv - seems like the tests are failing due to only one cluster being allowed. is this a service limit that can be increased or?

```
Test Failed

------- Stdout: -------
=== RUN   TestAccOpenShiftCluster_private
=== PAUSE TestAccOpenShiftCluster_private
=== CONT  TestAccOpenShiftCluster_private
=== CONT  TestAccOpenShiftCluster_private
    testcase.go:110: Step 1/1 error: Error running apply: exit status 1
        
        Error: creating Red Hat OpenShift Cluster "acctestaro220929200149494479" (Resource Group "acctestRG-aro-220929200149494479"): redhatopenshift.OpenShiftClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="DuplicateResourceGroup" Message="The provided resource group '/subscriptions/*******/resourcegroups/aro-cyufr257' already contains a cluster."
        
          with azurerm_redhatopenshift_cluster.test,
          on terraform_plugin_test.tf line 52, in resource "azurerm_redhatopenshift_cluster" "test":
          52: resource "azurerm_redhatopenshift_cluster" "test" {
        
--- FAIL: TestAccOpenShiftCluster_private (185.03s)
```

fedeoliv commented:
> s like the tests are failing due to only one cluster being allowed. is this a service limit that can be increased or

Hey @katbyte, humm seems this is a limitation. How did you get this error? Did it happen due to parallel tests execution? If yes, that's weird...I thought the random string value associated with a resource group name (e.g. accTestRG-aro-%d) wouldn't be a problem in terms of potential name conflicts.

david-yu commented:
@fedeoliv Any idea if we're able to resolve the test failures so we can get this PR merged?

david-yu commented:
@fedeoliv From chatting internally with some HashiCorp folks, it looks like a few clean up tasks are needed

> go mod tidy go mod vendor should fix it up - the real issue here is the failing tests due to that random resource group conflict

katbyte commented:
I am still seeing three distinct test failures for this PR:
<img width="1631" alt="image" src="https://user-images.githubusercontent.com/1638467/197260946-de632e2b-247b-4bff-bb78-a3ce5bb0cdc3.png">

It seems like for some of the tests the SP needs additional permissions? could we update the tests and document that?

david-yu commented:
Hi @fedeoliv do you know more help here to get this one through?

david-yu commented:
@fedeoliv Pinging again, looks like some tests are still failing, did you need help understanding why these failures are occurring? Does OpenShift require a dedicated RG for testing on Azure? We currently use a shared RG here for tests based on chatting with the team.

katbyte commented:
@david-yu - is there a resource group behind the scenes used by the service? because we seem to be passing in different one for each test and not defining "aro-xxxxx" anywhere in code?

david-yu commented:
You need to define a resource group ahead of time before you provision an Azure RedHat Openshift cluster as per these docs: https://learn.microsoft.com/en-us/azure/openshift/tutorial-create-cluster. I typically create one myself manually using the CLI.

katbyte commented:
@david-yu - we are:
<img width="720" alt="image" src="https://user-images.githubusercontent.com/1638467/214961450-eb82dd3e-7900-41ee-880e-a4098f9b0be5.png">

we create a rg `acctest-aro-$random`

while the error is "there is more then 1 cluster in rg aro-cyufr257" - a resource group we are not creating

david-yu commented:
@fedeoliv Do you have any idea where the extraneous resource group is coming from?

david-yu commented:
I'm not familiar with this code but it looks like it is coming from https://github.com/hashicorp/terraform-provider-azurerm/blob/2be494170676f293f53645e088c4fb4261475568/internal/services/redhatopenshift/redhatopenshift_cluster_resource.go#L720 for code to deal with a Openshift cluster with a custom domain.

katbyte commented:
That would do it @david-yu - is there any reason that cluster profile resource group would be different from the resource group for the resource?

david-yu commented:
I'm not sure why that cluster profile is different, I think we should just re-use the `resource_group_name` or if that breaks things perhaps expose that as a variable as you suggested.

yukirii commented:
When creating ARO cluster, two resource groups will be created.

This first group contains the resource of ARO cluster itself (`Microsoft.ContainerService/openShiftManagedClusters`). The group name is based on `azure_redhatopenshift_cluster.xxx.resource_group_name`.

The second resource group (`aro-*****`) is created by ARO resource prvider when creating cluster. This group contains all of the infrastructure resources associated with the ARO cluster such as VMs, networks, disks, etc.
(It is similar with [Node Resource Group (MC_*) of AKS](https://learn.microsoft.com/en-us/azure/aks/faq#why-are-two-resource-groups-created-with-aks).)

The name of second resource group is based on `clusterProfile.domain` property.

For example, Azure CLI generates `random_id` and use it as domain value by default.
If user not specified `--domain` to CLI option, `random_id` will be used for domain property. The second resource group will be created with a name which is contains domain (`random_id`).
https://github.com/Azure/azure-cli/blob/c3e19b1df5df4d1a9475bbda441a9fb4b284e308/src/azure-cli/azure/cli/command_modules/aro/custom.py#L101

![image](https://user-images.githubusercontent.com/1394098/215035500-d2c29049-2262-45aa-a9b7-ae779ed651a2.png)
![image](https://user-images.githubusercontent.com/1394098/215035481-5c6db8a1-09fa-4630-99bc-492cd5abcb20.png)

https://github.com/hashicorp/terraform-provider-azurerm/pull/15262#issuecomment-1405985601
If Terraform test cases running in parallel and each case are using same domain value, it may cause of conflict of resource group.

katbyte commented:
@yukirii - thanks for the context that helps. there are definitely some bugs in the PR then as it is not creating a new random domain for each creational call and sharing it for the duration of the provider instance - so that'll need to change so multiple can be created (or we can make it required/default to something else) 

do we need to pass in the resource group name during creation if the service automatically creates it from the domain? or will it just take whatever resourcegroup name passed in and automatically create whatever it is for the resources resource group? does it just use the resource's location?

Also i presume this resource group/etc will be all cleaned up during the cluster delete call?

guptaaayush8 commented:
I use terraform azapi( [Microsoft.RedHatOpenShift/openShiftClusters@2022-09-04](https://learn.microsoft.com/en-us/azure/templates/microsoft.redhatopenshift/openshiftclusters?pivots=deployment-language-terraform) ) to create the cluster below are some of the observations:
- Cluster resource group is created as a part of cluster; however, we have to pass it as id and not just name. (eg: /subscriptions/xxx-xxx-xxxx-xxxx/resourceGroups/terraformclusterauxrg) (any resource group name will work), and it gets deleted when cluster gets deleted.
- idempotency of the resource is doubtful as I am getting below error when changing worker count, it takes input as total worker node count and but expands those in an array.
![image](https://user-images.githubusercontent.com/99785775/215104152-b78fc6ea-1115-421f-b5c6-b5a8bc1dae0b.png)
- the name of the worker node is randomized when worker name is given as worker it randomizes it to <clustername>-x9gkk-worker-eastus2

david-yu commented:
@yukirii and @guptaaayush8 Could you help us test the changes and also continue to help answer some of the questions that @katbyte has to get this PR over the line and merged? Appreciate you for responding to @katbyte's questions.

katbyte commented on internal/services/redhatopenshift/redhatopenshift_cluster_resource.go:
```diff
@@ -0,0 +1,889 @@
+	input []interface{},
+	subscriptionId string,
+) *redhatopenshift.ClusterProfile {
+	resourceGroupName := fmt.Sprintf("aro-%s", randomDomainName)
```
we should make this user configurable, maybe required, and default it to something a bit more clear:
`$clusterrgname+"-resources"` or `redhatopenshift-$name-resources`

guptaaayush8 replied:
i suggest having it simply as a variable with no prefix and suffix, some users' needs to have resource group name in a specific fashion (if possible, with ability to add some tags)

katbyte replied:
what i'm suggesting is that we allow people to configure this by adding it to the cluster profile block, right now we force it to be aro-abcdef which is not ideal.  So lets make it configurable and then either make it required (what i would recommend) or default to something more clear

katbyte commented on internal/services/redhatopenshift/redhatopenshift_cluster_resource.go:
```diff
@@ -0,0 +1,889 @@
+	subscriptionId string,
+) *redhatopenshift.ClusterProfile {
+	resourceGroupName := fmt.Sprintf("aro-%s", randomDomainName)
+	resourceGroupId := ResourceGroupID(subscriptionId, resourceGroupName)
```
you can likely use the existing erg id functions here instead of a helper

katbyte commented on internal/services/redhatopenshift/redhatopenshift_cluster_resource.go:
```diff
@@ -0,0 +1,889 @@
+
+	pullSecret := config["pull_secret"].(string)
+
+	domain := config["domain"].(string)
+	if domain == "" {
+		domain = randomDomainName
+	}
```
this should be set in the create where it is randomly created if the user doesn't specify it not here as this is shared with update. and is there an argument to just make it required? or is the random string how the portal/service behaves?

guptaaayush8 replied:
it is mandatory field in azure portal.

katbyte replied:
ah well then if its mandatory in the portal we should just make it mandatory in terraform too and ensure the user picks the name rather then randomly generating it

katbyte reviewed (changes requested):
Left some more detailed comments inline

mbfrahry commented:
Hey all! I've taken this over and am reviewing/rewriting this on #20266 so I'll be closing this PR down. Thanks for getting this started @fedeoliv!
